spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+  upon `PrepareProposal` or `ProcessProposal` and the resulting state was kept in memory, the
+  Application can simply apply that state (faster) to the main state, rather than reexecuting
+  the decided block (slower).
+
+`PrepareProposal`/`ProcessProposal` can be called many times for a given block height. Moreover,
+it is impossible to accurately predict which of the proposed blocks for a height will finally make
+it into that height's `FinalizeBlock`.
+Therefore, the state resulting from exectuing proposed blocks, denoted *candidate state*, needs
+to be kept in memory as an alternative. When `FinalizeBlock` is called, the Application can
+check if the decided block corresponds to one of its candidate states; if so, it will apply it as
+its `ExecuteTxState` (see [Consensus Connection](#consensus-connection) below),
+which will be persisted during the upcoming `Commit` call.
+
+In particularly bad runs (e.g., because of network instability), Consensus might take many rounds.
+In those cases many proposed blocks will be disclosed to the Application for a given height.
+Indeed, by the nature of Consensus, the number of proposed blocks received by the Application for
+a particular height cannot be bound, so Application developers must act with care and use mechanisms
+to bound memory usage. As a general rule, the Application should be ready to discard candidate states
+before `FinalizeBlock`, even if one of them might end up corresponding to the
+decided block and thus have to be reexecuted upon `FinalizeBlock`.
+
+### States and ABCI++ Connections
+
+#### Consensus Connection
+
+The Consensus Connection should maintain an `ExecuteTxState` - the working state
+for block execution. It should be updated by the call to `FinalizeBlock`
+during block execution and committed to disk as the "latest
+committed state" during `Commit`. Execution of a proposed block (via `PrepareProposal`/`ProcessProposal`) must not update the `ExecuteTxState`, but rather be kept as a separate candidate state until `FinalizeBlock`
+confirms which of the candidate states (if any) can be used to update `ExecuteTxState`.
+
+#### Mempool Connection
+
+The mempool Connection should maintain a `CheckTxState`
+to sequentially process pending transactions in the mempool that have
+not yet been committed. It should be initialized to the latest committed state
+at the end of every `Commit`.
+
+During the execution of a Consensus instance, the `CheckTxState` may be updated concurrently with the
+`ExecuteTxState`, as messages may be sent concurrently on the Consensus and Mempool connections.
+At the end of the Consensus instance, as described above, Tendermint will lock the mempool and flush
+the mempool connection before calling `Commit`, ensuring that all existing `CheckTx` calls are
+responded to and no new ones can begin.
+
+After `Commit`, while still holding the mempool lock, `CheckTx` is run again on all transactions
```

Comment:
> This is not mandatory though, as it can be disabled by a config flag.
Yes but do we really want to include that detail here?

> After the commit call returns, is that what you mean?
Yes. Will reword